src/coreclr/jit/morph.cpp

```diff
@@ -2043,17 +2043,23 @@ void CallArgs::AddFinalArgsAndDetermineABIInfo(Compiler* comp, GenTreeCall* call
         PushBack(comp, NewCallArg::Primitive(newArg).WellKnown(WellKnownArg::WrapperDelegateCell));
     }
 #endif // defined(TARGET_ARM)
-#ifndef TARGET_X86
+
+    bool addStubCellArg = true;
+
+#ifdef TARGET_X86
     // TODO-X86-CQ: Currently RyuJIT/x86 passes args on the stack, so this is not needed.
     // If/when we change that, the following code needs to be changed to correctly support the (TBD) managed calling
     // convention for x86/SSE.
 
+    addStubCellArg = call->gtCallType != CT_INDIRECT && comp->IsTargetAbi(CORINFO_NATIVEAOT_ABI);
```

Comment:
How are the indirect cases handled on NAOT? Does it always assume the stub cell is in `eax`? Because then it probably works only by chance, and this code should just be

```suggestion
    addStubCellArg = comp->IsTargetAbi(CORINFO_NATIVEAOT_ABI);
```

Reply:
`CT_INDIRECT` takes a different code path (just like in R2R and JIT) which explicitly set the EAX. It needs to be skipped here otherwise we would try to allocate EAX twice and hit an assert.

https://github.com/dotnet/runtime/blob/0cb48f4f2de372118a324be32a3de77340c30f22/src/coreclr/jit/codegenxarch.cpp#L6281-L6320
https://github.com/dotnet/runtime/blob/0cb48f4f2de372118a324be32a3de77340c30f22/src/coreclr/jit/lsraxarch.cpp#L1332-L1343
https://github.com/dotnet/runtime/blob/0cb48f4f2de372118a324be32a3de77340c30f22/src/coreclr/jit/lowerxarch.cpp#L6637-L6651